makefile

```diff
@@ -12,15 +12,35 @@ $(1)
 
 endef
 
+# Function to make sure that the git directory is clean.
+define check-git-status
+if [[ -n $$(git --no-pager status -s 2> /dev/null) ]] ;\
+then \
+	echo "Git tree is not clean. Did you forget to commit some files?" ;\
+	git --no-pager status ;\
+fi
+endef
+
 # CI build
 .PHONY: ci
-ci: gen-java gen-swagger
+ci: validate-go gen-java gen-swagger
+
+# Validate that generated Go files are up to date.
+.PHONY: validate-go
+validate-go: gen-go-ci
+	$(check-git-status)
+
+# Generate ProtoBuf implementation for Go.
+.PHONY: gen-go-ci
+gen-go-ci:
+	rm -rf ./$(GENDIR)/go
+	$(foreach file,$(PROTO_FILES),$(call exec-command,protoc --go_out=plugins=grpc:$(GOPATH)/src $(file)))
 
 # Generate ProtoBuf implementation for Go.
 .PHONY: gen-go
 gen-go:
-	$(foreach file,$(PROTO_FILES),$(call exec-command,protoc --go_out=plugins=grpc:$(GOPATH)/src $(file)))
 	rm -rf ./$(GENDIR)/go
+	$(foreach file,$(PROTO_FILES),$(call exec-command,protoc --go_out=plugins=grpc:$(GOPATH)/src $(file)))
```

Comment:
This seems identical to lines 36-37 (I cannot see a difference if there is). If so why not `gen-go: gen-go-ci` or some other way to reuse it?

Reply:
The difference between `gen-go` and `gen-go-ci` is that apparently in Travis we check out sources into GOPATH, so `cp` fails to copy from directory to itself.

Agreed, need to deduplicate this.